lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,470 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This charm library contains utilities to automatically forward your charm logs to a loki-push-api endpoint.
+
+(yes! charm code, not workload code!)
+
+This means that, if your charm is related to, for example, COS' Loki charm (or a Grafana Agent),
+you will be able to inspect in real time from the Grafana dashboard the logs emitted by your charm.
+
+To start using this library, you need to do two things:
+1) decorate your charm class with
+
+`@log_charm(loki_push_api_endpoint="my_logging_endpoint")`
+
+2) add to your charm a "my_logging_endpoint" (you can name this attribute whatever you like) **property**
+that returns an http/https endpoint url. If you are using the `LogProxyConsumer` as
+`self.logging = LogProxyConsumer(self, ...)`, the implementation could be:
+
+```
+    @property
+    def my_logging_endpoint(self) -> List[str]:
+        '''Loki push API endpoints for charm logging'''
+        return self.logging.loki_endpoints:
+```
```

Comment:
Can we pass a relation name in instead of the endpoints? Then just get the endpoints from relation data. Seems like a better solution to me.

Reply:
We made the same decision for charm tracing. It's a matter of separation of concerns: we don't want the charm logging lib to make assumptions about the relation interface. Also take a look at loki's own usage: in loki's case, we don't get the endpoint(s) from relation data, we simply return `"http://localhost:3100/api/v1/push"`.

The library shouldn't have to worry about databag parsing.

Reply:
another example: we'd also have to automatically detect if the relation you're passing is to a grafana-agent instead of a loki, and fetch the push api endpoint in a different way.

Reply:
I don't understand. We know exactly what the interface looks like. It is always loki_push_api. Grafana-agent uses the exact same interface. We don't need to care at all whether it is grafana-agent or loki on the other end.

Reply:
resolved in face to face convo, agreed that the charm lib can and should remain interface agnostic